### PR TITLE
Making ICEs and test them in incremental

### PR DESCRIPTION
**Pull request by traxys**
This adds:
 - A way to make the compiler ICE
 - A way to check for ICE in `cfail` tests with `should-ice`
 - A regression test for issue #65401

I am not sure the attribute added `should-ice` is the best for this job

### CODE REVIEW

**Review by michaelwoerister — COMMENTED** [review]
Thanks for the PR. @traxys! This looks good in general. I wonder though if the `// should-ice` functionality can be achieved with existing means. I'll take a closer look later today. Meanwhile, would you mind addressing the nit listed below?

**Comment by michaelwoerister** [comment]
@traxys, I'm not quite clear if `// should-ice` brings new functionality. Could you also write the tests as the following?
```rust
// revisions: cfail1 cfail2
// error-pattern: delayed span bug triggered by #[rustc_error(delay_span_bug_from_inside_query)]
// failure-status: 101

#![feature(rustc_attrs)]

#[rustc_error(delay_span_bug_from_inside_query)]
fn main() {}
```

**Comment by traxys** [comment]
@michaelwoerister the problem is this line https://github.com/rust-lang/rust/blob/3a1b3b30c6cdd674049b144a3ced7b711de962b2/src/tools/compiletest/src/runtest.rs#L1407, there is no check to avoid it

**Comment by michaelwoerister** [comment]
> There is currently no way to avoid the early stop

OK, that makes sense. In that case adding `// should-ice` looks like the way to go.

File: src/tools/compiletest/src/runtest.rs [hunk 1]
```diff
@@ -428,7 +434,12 @@ impl<'test> TestCx<'test> {
     }
 
     fn check_correct_failure_status(&self, proc_res: &ProcRes) {
-        let expected_status = Some(self.props.failure_status);
+        let expected_status =
+            if self.props.should_ice {
+                Some(101)
+            } else {
+                Some(self.props.failure_status)
+            };
```
Comment (michaelwoerister) [inline src/tools/compiletest/src/runtest.rs]: Would you mind moving this adaptation of the failure status to `header.rs` where we already do something similar:
https://github.com/rust-lang/rust/blob/e4931eaaa3d95189b30e90d3af9f0db17c41bbb0/src/tools/compiletest/src/header.rs#L574-L579

Then it's all in one place.

**Comment by traxys** [comment]
I think the `should-ice` should be restricted to `cfail` tests, meaning if it is not a `cfail` test the `compiletest` errors. Is it a good idea ?

**Comment by michaelwoerister** [comment]
Yeah, that's fine.

**Comment by traxys** [comment]
I don't know if I put the test in the best place, but it should now error on misuse of `should-ice`

**Comment by michaelwoerister** [comment]
Thanks, @traxys! This looks good to me now.

@bors r+

**Pull request merged**